pjsip/src/pjsua-lib/pjsua_call.c

```diff
@@ -5204,6 +5204,7 @@ static void pjsua_call_on_media_update(pjsip_inv_session *inv,
 	 * UAS dialog and if this is not a re-INVITE
 	 */
 	if (inv->state != PJSIP_INV_STATE_NULL &&
+	    inv->state != PJSIP_INV_STATE_EARLY &&
```

Comment:
This patch seems to cause a behavior change, i.e: if the SDP offer/answer nego from INVITE transaction (not UPDATE) fails, the call will be no longer disconnected, or even the call will never get disconnected on any SDP nego failure.

As the reported issue seems to be SDP nego failure of UPDATE while processing the initial INVITE, the fix should affect that specific scenario only. Unfortunately the callback does not provide `tsx` or `rdata` info, so we cannot check if it is SDP offer/answer from UPDATE, but perhaps we can use `call->med_cnt` info for checking whether there has been a successful SDP nego before (from INVITE+18x)?

Reply:
Hi @nanangizz , this issue was found, trying to fix #2967, so yes it should affect not only UPDATE, but also PRACK SDP offer/answer negotiation.
I tried, but I really can't imagine the scenario, when the call in EARLY state must be disconnected. Can you please describe the scenario? Then I can prepare the sipp test, that is failing and fix it.

Reply:
AFAIK, the current behavior of PJSUA lib is as follow:

For UAC:
1. send INVITE with SDP offer
2. recv 183 with SDP answer
3. when UAC sees failure in SDP nego, it will disconnect call (by sending CANCEL)

For UAS:
1. receive INVITE with SDP offer
2. generate SDP answer (e.g: for 183 or 200 response)
3. the generated answer will be negotiated with the offer, if it fails, call will be disconnected

The proposed patch seems to change this behavior?

Reply:
I prepared UAC and UAS tests with 3 different unacceptable SDP:
1) audio codec that is not supported by pjsua 
2) media type video, instead of audio
3) SDP version=1 (must be v=0)
Don't be confused by test names, they show the role of sipp (uac/uas)

First of all, I can confirm, that this patch changes behavior in one case (pjsua is UAC and receives 183 with SDP media type=video) uas-answer-183-wrong-media.xml
It can be fixed for example with this not too elegant code:
```
 if ((inv->state != PJSIP_INV_STATE_NULL &&
    inv->state != PJSIP_INV_STATE_EARLY &&
    inv->state != PJSIP_INV_STATE_CONFIRMED) ||
    (inv->state == PJSIP_INV_STATE_EARLY && call->med_cnt == 0))
```
Interesting case is number 3. Pjsua receives completely wrong SDP. It does not work correctly in both UAC/UAS scenarios even without any patches, but using current source code.
When pjsua is UAC (uas-answer-183-wrong-version.xml) it detects an error `Error parsing/validating SDP body: Invalid SDP version line`, but never disconnects the call.
When pjsua is UAS (uac-183-wrong-version.xml) it answers `400 Bad SDP`, receives ACK, but continues retransmitting "400 Bad SDP". Call is not disconnected too.
Function `inv_on_state_early` never checks return status of `inv_check_sdp_in_incoming_msg`.
Tests are attached.
[uas-answer-183-wrong-version.xml.txt](https://github.com/pjsip/pjproject/files/9828542/uas-answer-183-wrong-version.xml.txt)
[uas-answer-183-wrong-version.py.txt](https://github.com/pjsip/pjproject/files/9828543/uas-answer-183-wrong-version.py.txt)
[uas-answer-183-wrong-media.xml.txt](https://github.com/pjsip/pjproject/files/9828544/uas-answer-183-wrong-media.xml.txt)
[uas-answer-183-wrong-media.py.txt](https://github.com/pjsip/pjproject/files/9828545/uas-answer-183-wrong-media.py.txt)
[uas-answer-183-wrong-codec.xml.txt](https://github.com/pjsip/pjproject/files/9828546/uas-answer-183-wrong-codec.xml.txt)
[uas-answer-183-wrong-codec.py.txt](https://github.com/pjsip/pjproject/files/9828547/uas-answer-183-wrong-codec.py.txt)
[uac-183-wrong-version.xml.txt](https://github.com/pjsip/pjproject/files/9828548/uac-183-wrong-version.xml.txt)
[uac-183-wrong-version.py.txt](https://github.com/pjsip/pjproject/files/9828549/uac-183-wrong-version.py.txt)
[uac-183-wrong-media.xml.txt](https://github.com/pjsip/pjproject/files/9828550/uac-183-wrong-media.xml.txt)
[uac-183-wrong-media.py.txt](https://github.com/pjsip/pjproject/files/9828551/uac-183-wrong-media.py.txt)
[uac-183-wrong-codec.xml.txt](https://github.com/pjsip/pjproject/files/9828552/uac-183-wrong-codec.xml.txt)
[uac-183-wrong-codec.py.txt](https://github.com/pjsip/pjproject/files/9828553/uac-183-wrong-codec.py.txt)

Reply:
Thanks for the pjsua/sipp tests! So the behavior change is confirmed and we need to fix it. And regarding case no 3, perhaps we should create a separate issue/PR? Good catch btw.

Reply:
> ```
>  if ((inv->state != PJSIP_INV_STATE_NULL &&
>     inv->state != PJSIP_INV_STATE_EARLY &&
>     inv->state != PJSIP_INV_STATE_CONFIRMED) ||
>     (inv->state == PJSIP_INV_STATE_EARLY && call->med_cnt == 0))
> ```

I'm wondering if checking `if (call->med_cnt == 0)` only actually is sufficient here? since we should only disconnect the call if there's no media, otherwise we should revert the previous media.

Reply:
@sauwming I'm completely newbie in pjproject code, so I'm not 100% sure, that `(call->med_cnt == 0)` covers all the cases, when call should be disconnected in EARLY state. It's worth to make tests with incompatible SDP for late negotiation for example. But they should also have `call->med_cnt == 0`
According to RFC 6337 section 2.2 (https://datatracker.ietf.org/doc/html/rfc6337#section-2.2) only PRACK and UPDATE can send SDP offer in EARLY state, but `call->med_cnt == 1` in these cases.
Actually, all these tests with wrong SDP in 183 are emulating evil UAS, because normal response must be 488 or something similar. There can be other unpredictable cases, which break RFCs.
@nanangizz I'll create one more issue for 183 reply with broken SDP (e.g. v=1)

Reply:
Also, adding some comments (and/or add link to this PR) about the new added checks is usually useful for future ref.